sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -42,10 +42,34 @@ case class InMemoryTableScanExec(
   override def output: Seq[Attribute] = attributes
 
   // The cached version does not change the outputPartitioning of the original SparkPlan.
-  override def outputPartitioning: Partitioning = relation.child.outputPartitioning
+  // But the cached version could alias output, so we need to replace output.
+  override def outputPartitioning: Partitioning = {
+    val attrMap = AttributeMap(
+      relation.child.output.zip(output)
+    )
+    relation.child.outputPartitioning match {
+      case HashPartitioning(expressions, numPartitions) =>
+        val newExprs = expressions.map(_.transform {
+          case attr: Attribute if attrMap.contains(attr) => attrMap.get(attr).get
+        })
+        HashPartitioning(newExprs, numPartitions)
+      case _ => relation.child.outputPartitioning
+    }
+  }
 
   // The cached version does not change the outputOrdering of the original SparkPlan.
-  override def outputOrdering: Seq[SortOrder] = relation.child.outputOrdering
+  // But the cached version could alias output, so we need to replace output.
+  override def outputOrdering: Seq[SortOrder] = {
+    val attrMap = AttributeMap(
+      relation.child.output.zip(output)
+    )
+    relation.child.outputOrdering.map { sortOrder =>
+      val newSortExpr = sortOrder.child.transform {
```

Comment:
we can even have a method `def updateAttribute(expr: Expression): Expression` to do this.